Add ledger_entry options for fee, amendments, NUNL, and hashes

ximinez opened the pull request:
## High Level Overview of Change

Adds `ledger_entry` functionality for global record-keeping objects, with some flexibility.

Also will always return the hash for any valid `ledger_entry` request, even if the object is not found or valid.

### Context of Change

While doing some analysis, I was annoyed that I couldn't look up these records without computing their indexes or finding them on the [ledger_entry](https://xrpl.org/docs/references/http-websocket-apis/public-api-methods/ledger-methods/ledger_entry#get-ledger-entry-by-id) documentation page.

### Type of Change

- [X] New feature (non-breaking change which adds functionality)
- [X] Tests (you added tests for code that already exists, or your new feature included in this PR)

### API Impact

- [X] Public API: New feature (new methods and/or new fields)

## Before / After

Affects the `Amendments`, `FeeSettings`, `NegativeUNL`, and `LedgerHashes` objects.

### Before:

These objects could only be found using, for example, `"index" : "7DB0788C020F02780A673DC74757F23823FA3014C1866E72CC4CD8B226CD6EF4"` for `Amendments`.

Additionally, if an object is not found, only the error is returned.

### After:

These objects can be found in multiple ways:

1. Using the format of other ledger objects, like accounts, but the only valid value is `true` or the valid index of the object: `"amendments" : true`. e.g. `"amendments" : true`.
	1. `Amendments`: `"amendments" : true`
	2. `FeeSettings`: `"fee" : true`
	3. `NegativeUNL`: `"nunl" : true`
	4. `LedgerHashes`: `"hashes" : true` (For the "short" list. See below.)
2. Using special case values to `"index"`, such as `"index" : "amendments"`. Uses the same names as above. Note that for `"hashes"`, this option will only return the recent ledger hashes / "short" skip list.
3. `LedgerHashes` has two types: "short", which stores the recent hashes since the last flag ledger, and "long", which stores the flag ledger hashes for a particular ledger range.
	1. To find a "long" `LedgerHashes` object, request `"hashes" : <ledger sequence>`. `<ledger sequence>` can be integer value that evaluates to an unsigned integer.
	2. To find the "short" `LedgerHashes` object, request `"hashes": true` as with the other fixed types.

The following queries are all functionally equivalent:
* `"amendments" : true`
* `"index" : "amendments"`
* `"amendments" : "7DB0788C020F02780A673DC74757F23823FA3014C1866E72CC4CD8B226CD6EF4"`
* `"index" : "7DB0788C020F02780A673DC74757F23823FA3014C1866E72CC4CD8B226CD6EF4"`

Finally, whether the object is found or not, if the query results in a valid index, that index will be returned. This can be used to confirm the query was valid, or to save the index for future use.

ximinez committed:
Add ledger_entry options for fee, amendments, NUNL, and hashes

- Request format for fee, amendments, and nunl is
  "fieldName":<any valid json>, including null. Because these types
  have no inputs to the index computation, any value
  works.
- Request format for hashes is "hashes":<uint value or any valid
  json>. If the value is numeric, and evaluates to a uint, return the
  hashes history page for that ledger sequence. (AKA "long" skip list.)
  Any other valid json value will return the "short" skip list.

ximinez committed:
Add support for "index":"fixed type" lookups in ledger_entry

- For example "index":"amendments" will return the amendments object as
  if the request was
  "index":"7DB0788C020F02780A673DC74757F23823FA3014C1866E72CC4CD8B226CD6EF4"
- Works for amendments, fee, nunl, and hashes (for the recent history
  list only).
- Always returns the valid computed index, even if the object is not
  found.

mvadari commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -41,8 +41,19 @@ namespace ripple {
 static std::optional<uint256>
 parseIndex(Json::Value const& params, Json::Value& jvResult)
 {
+    std::string const index = params.asString();
```
I think I may have miscommunicated my suggestion - I was suggesting that 
`amendments: "01234...."` should only accept the actual `amendments` key, and not other values.

ximinez replied:
Only the correct key will give you the `Amendments` object. Anything else will give you `objectNotFound` or `unexpectedLedgerType`.

mvadari replied:
I disagree with supporting these strings. Currently `index` only supports uint256s and nothing else, and this feels like a breaking change to that API to me.

ximinez replied:
Yes, I see your point. I have counterpoints
1. _Technically_, `index` only supports strings, which are then parsed into `uint256`. Since we're parsing anyway, why not parse a few special-case strings into other `uint256`s?
2. I was modelling the behavior on the `ledger_index` parameter, which normally takes numbers, but can also take the special-case values of "open", "closed", and "validated".
3. I usually think of breaking changes as things that move or remove an input or output value, such that an existing integration that's expecting something to be in a certain place can't find it anymore. This is more like an addition. It can be completely ignored, and nothing will suffer for it.

I'm willing to take it out or gate it on API version if you still think it's breaking. Thoughts?

ximinez committed:
Refactor a little bit to reduce future conflicts

- Specifically #5237

ximinez commented:
Thank you for explaining!

My main goal with this change, aside from just getting something I could use myself, was to make it as easy as possible to get the information I wanted. I avoided doing any validation of the input value, because it makes absolutely no difference to the lookup functionality.

> > * Should the value be truthy?
> 
> Instead of accepting all `bool`s provided into the `amendments` field, only accept `true`. Or don't accept `0` or negative numbers, only positive numbers. I thought the concept of "truthiness" was a general programming thing, but it seems to only be in JS: [geeksforgeeks.org/javascript/explain-the-concept-of-truthy-falsy-values-in-javascript](https://www.geeksforgeeks.org/javascript/explain-the-concept-of-truthy-falsy-values-in-javascript/)
> 
> Edit after reading the description: I think passing in `null` and expecting not-an-error to be returned is rather confusing.

Ah. I see. I was loosely familiar with "truthiness", but I didn't see how it applied here, since I didn't care about the value, only they key. That was partially informed by the implementation, where the first matched key is the one used for the query. To me it would be nonsensical to request `{ "account_root" : false, "check" : "ABCD..." }`, because it'll only process the first hit (`account_root`), which will fail, because `false` is an invalid input. Further, it would be absurd to expect a `Check` in the result.

Given that, and given that the input doesn't need to be parsed, and given that I want it to be easy, I opted to do no validation of any kind.

It could be added, but my attitude is that nobody is going to complain about _not_ getting an error, right?

> > * If a string hash is provided, shouldn't that be checked for correctness?
> 
> You can provide `amendments: "01234..."`. Shouldn't that string be required to only be the actual fixed key, rather than accepting any value?

Ah. I understand. This question basically has the same answer as the first. Aside from the fact that passing a key would be silly, since you can just use `"index":"01234..."`, if I'm not doing any validation of the input, then they just get the `Amendment` object back, and if that's not what they expect, they need to write a better request.

ximinez commented:
> my attitude is that nobody is going to complain about _not_ getting an error, right?

On second thought. I forgot my old maxim that "a wrong answer is worse than no answer". And if someone passes some other index as the value for `"amendments":value`, expecting the entry at that index, then returning the `Amendments` object is, arguably, a wrong answer.

I'm rewriting it now so that `true` is the only valid value for all these queries, with the addition of a "ledger indexy" value for `hashes`.

mvadari commented:
> > my attitude is that nobody is going to complain about _not_ getting an error, right?
> 
> On second thought. I forgot my old maxim that "a wrong answer is worse than no answer". And if someone passes some other index as the value for `"amendments":value`, expecting the entry at that index, then returning the `Amendments` object is, arguably, a wrong answer.
> 
> I'm rewriting it now so that `true` is the only valid value for all these queries, with the addition of a "ledger indexy" value for `hashes`.
> 
> _Edit:_ `true` or an index key.

I would perhaps accept all non-null, non-string values (plus the actual key string). `null` because I hate how `Json::Value` treats `null` as a real value and I don't want to support it here as a real value, especially since `Json::Value` will automatically add a null value if you do `obj[key]` before checking membership.

ximinez committed:
Restrict which values are valid for fixed types.

- True or a ledger index for those that can't parameters
- Number, true, or a ledger index for "hashes".

ximinez commented:
> I would perhaps accept all non-null, non-string values (plus the actual key string). `null` because I hate how `Json::Value` treats `null` as a real value and I don't want to support it here as a real value, especially since `Json::Value` will automatically add a null value if you do `obj[key]` before checking membership.

I decided to limit it to `true` or the key string for simplicity, both for the user, and for documentation. Also because a number has meaning to `hashes`, but it otherwise follows the same rules, so this gives consistency.

tequdev commented:
How about supporting `amendment: {}`? 
This would be consistent with other non-fixed types.

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez commented:
> How about supporting `amendment: {}`? This would be consistent with other non-fixed types.

It wouldn't be all that consistent, IMHO. First json objects are used to pass additional parameters. There are no additional parameters needs here. Second, not all of the other types accept objects. `AccountRoot` and `Check` are two examples

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into ximinez/fixed-ledger-entries

* upstream/develop:
  Refactor `ledger_entry` RPC source code and tests (5237)

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Gate the new "index: special_case" on the current beta apiVersion 3

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Move the lambda to the top so that the parse functions stay alphabetical

mvadari commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -692,6 +792,23 @@ struct LedgerEntry
+// Helper function to return FunctionType for objects that have a fixed
+// location. That is, they don't take parameters to compute the index.
+// e.g. amendments, fees, negative UNL, etc.
+static FunctionType
```
I think this is nicer, but if you prefer the old version that's fine with me.

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

mvadari commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -54,13 +80,33 @@ parseObjectID(
+    Json::StaticString const fieldName,
+    unsigned apiVersion)
+{
+    if (apiVersion > 2u && params.isString())
```
Do we have a catalog of what is going into APIv3 yet? I don't want this to fall through the cracks when it's time to get feedback on/release APIv3.

mvadari replied:
cc @bthomee

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Review feedback from @mvadari: Document test lambdas

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

mvadari commented on src/test/rpc/LedgerEntry_test.cpp:
```diff
@@ -2027,6 +2029,370 @@ class LedgerEntry_test : public beast::unit_test::suite
+            checkResult(good, jv, expectedType);
+            BEAST_EXPECT(jv[jss::result][jss::index].asString() == hexKey);
+
+            // Use the "index":"field" notation with API 3
```
There should be a test to ensure this doesn't work with API v2

ximinez replied:
> There should be a test to ensure this doesn't work with API v2

Added in b165fd0514

mvadari reviewed (approved):
A couple of minor comments but nothing to block approval

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge remote-tracking branch 'XRPLF/develop' into ximinez/fixed-ledger-entries

* XRPLF/develop:
  Implement Lending Protocol (unsupported) (5270)

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Update View info() to header()

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge remote-tracking branch 'XRPLF/develop' into ximinez/fixed-ledger-entries

* XRPLF/develop:
  test: add more tests for `ledger_entry` RPC (5858)
  refactor: Rename `rippled.cfg` to `xrpld.cfg` (6098)
  Revert "chore: Pin ruamel.yaml<0.19 in pre-commit-hooks (6166)" (6167)
  chore: Pin ruamel.yaml<0.19 in pre-commit-hooks (6166)
  fix: Remove cryptographic libs from libxrpl Conan package (6163)

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

pratikmankawde commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -18,6 +18,32 @@
+static FunctionType
+fixed(Keylet const& keylet)
+{
+    return [&keylet](
```
Capturing `keylet` by reference here might cause issues if lifetime of passed `Keylet` object is limited to the call to `fixed()` or in any way not guaranteed to surpass the point where the returned lambda is called. Consider following call:
```cpp
auto returnedLambda = fixed(Keylet()); // or fixed({<parameters passed to the Keylet constructor>});
returnedLambda(); // This might fail or not give correct results, since at this point the Keylet object was destroyed.

```

pratikmankawde replied:
Although in current scenario, the `Keylet` objects retrieved using `keylet::*` and passed to `fixed(..)` are returning static objects. So should be fine for now.

ximinez replied:
> Although in current scenario, the `Keylet` objects retrieved using `keylet::*` and passed to `fixed(..)` are returning static objects. So should be fine for now.

I don't want to have to rely on these parameters being `static` forever, and this is RPC code that isn't highly performance-sensitive, so I think it'll be safer in the long run to make the copy.

Fixed in 41d222c227

pratikmankawde commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -33,13 +59,33 @@ parseObjectID(
+parseIndex(
+    Json::Value const& params,
+    Json::StaticString const fieldName,
+    unsigned apiVersion)
```
Minor: `apiVersion` can be `const`. Same comment for other functions where `apiVersion` is being passed.
```cpp
const unsigned apiVersion
```

ximinez replied:
> Minor: `apiVersion` can be `const`. Same comment for other functions where `apiVersion` is being passed.
> 
> ```c++
> const unsigned apiVersion
> ```

`const` isn't super important for non-reference function parameters, because there are no external consequences if the function changes it, but I went ahead and added it in 41d222c227

pratikmankawde replied:
Yes true. Sometimes though a developer can change the value of the passed in param (intentionally or un-intentionally when local variables have similar name), and then(same or another dev.) mistakenly assume at a later stage that the variable still holds original passed value. Making this `const` reduces chances of such issues.

pratikmankawde commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -85,7 +130,10 @@ parseAMM(Json::Value const& params, Json::StaticString const fieldName)
+parseBridge(
+    Json::Value const& params,
+    Json::StaticString const fieldName,
+    unsigned apiVersion)
```
Minor: `apiVersion` could be tagged with `[[maybe_unused]]` incase we activate the unused-param warning in the future.

ximinez replied:
> Minor: `apiVersion` could be tagged with `[[maybe_unused]]` incase we activate the unused-param warning in the future.

Added in 41d222c227

pratikmankawde reviewed (commented):
Left a-few comments.

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Review feedback from @pratikmankawde:

- Variable lifetime correctness
- const correctness and [[maybe_unused]] for apiVersion

ximinez committed:
Review feedback from @mvadari: Tests

- Scope `params` instead of clearing it between each test.
- Add a test to ensure "index: field" does not work with API < 3

ximinez commented on src/test/rpc/LedgerEntry_test.cpp:
```diff
@@ -2027,6 +2029,370 @@ class LedgerEntry_test : public beast::unit_test::suite
+            BEAST_EXPECT(!jv[jss::result].isMember(jss::index));
+
+            // "field":"string"
+            params.clear();
```
> Why clear `params` instead of scoping `params` in a `{}` block?

I honestly don't remember. I've changed it in b165fd0514

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

pratikmankawde commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -384,20 +451,53 @@ parseEscrow(Json::Value const& params, Json::StaticString const fieldName)
+parseFixed(
+    Keylet const& keylet,
+    Json::Value const& params,
+    Json::StaticString const& fieldName,
```
```suggestion
   [[maybe_unused]] Json::StaticString const& fieldName,
```
When params is `bool` type, `fieldName` will be unused.

ximinez replied:
> When params is `bool` type, `fieldName` will be unused.

I think the `maybe_unused` attribute only applies at compile time, not run time. Since it's referenced in the function, it's used, regardless of the parameter at runtime.

pratikmankawde reviewed (approved):
Looks good.

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries

ximinez committed:
Merge branch 'develop' into ximinez/fixed-ledger-entries
